wodles/azure/tests/test_azure.py

```diff
@@ -0,0 +1,859 @@
+    mock_datetime.return_value = parse(desired_date)
+    azure.args = MagicMock(reparse=reparse, graph_query=query)
+
+    result = azure.build_graph_url(offset=offset, md5_hash=md5_hash)
+
+    mock_get.assert_called_with(azure.orm.Graph, md5=md5_hash)
+    mock_create.assert_called_with(table=azure.orm.Graph, query=query, md5_hash=md5_hash, offset=offset)
+
+    filtering_condition = "createdDateTime" if "signins" in query.lower() else "activityDateTime"
+
+    if reparse:
+        expected_str = f"{filtering_condition}+ge+{desired_date}"
+    else:
+        if parse(desired_date) < parse(min_date, fuzzy=True):
+            expected_str = f"({filtering_condition}+lt+{min_date}+and+{filtering_condition}+ge+{desired_date})" \
+                           f"+or+({filtering_condition}+gt+{max_date})"
+        elif parse(desired_date) > parse(max_date, fuzzy=True):
+            expected_str = f"{filtering_condition}+ge+{desired_date}"
+        else:
+            expected_str = f"{filtering_condition}+gt+{max_date}"
+    assert expected_str in result
+
+
+@patch('azure-logs.logging.error')
+@patch('azure-logs.orm.get_row', side_effect=azure.orm.AzureORMError)
+def test_build_graph_url_ko(mock_get, mock_logging):
+    """Test build_log_analytics_query handles ORM exceptions."""
+    with pytest.raises(SystemExit) as err:
+        azure.build_graph_url(offset=None, md5_hash=None)
+    assert err.value.code == 1
+    mock_logging.assert_called_once()
+
+
+@patch('azure-logs.send_message')
+@patch('azure-logs.update_row_object')
+@patch('azure-logs.get')
+def test_get_graph_events(mock_get, mock_update, mock_send):
+    """Test get_graph_events recursively request the data using the specified url and process the values present in the
+    response."""
+    def load_events(path):
+        with open(os.path.join(TEST_DATA_PATH, path)) as f:
+            return json.loads(f.read())
+
+    azure.args = MagicMock(graph_query="test_query", graph_tag="tag")
+
+    # The first file contains both values and a nextLink to the following file
+    event_list = MagicMock(status_code=200)
+    contents = load_events("graph_events")
+    event_list.json.return_value = contents
+    num_events = len(contents['value'])
+    url = contents['@odata.nextLink']
+
+    # The second file has no values nor nextLink
+    empty_event_list = MagicMock(status_code=200)
+    empty_event_list.json.return_value = load_events("graph_events_no_values")
+
+    mock_get.side_effect = [event_list, empty_event_list]
+
+    headers = "headers"
+    azure.get_graph_events(url=url, headers=headers, md5_hash="")
+    mock_get.assert_called_with(url=url, headers=headers)
+    assert mock_update.call_count == num_events
+    assert mock_send.call_count == num_events
+
+
+@pytest.mark.parametrize('status_code', [400, 500])
+@patch('azure-logs.logging.error')
+@patch('azure-logs.get')
+def test_get_graph_events_ko(mock_get, mock_logging, status_code):
```

Comment:
Not testing exceptions, change the test name or include these cases in the last test.